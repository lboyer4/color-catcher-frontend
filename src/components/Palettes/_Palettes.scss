.palette-holder {
	background-color: #F9F4F5;
	width: 100%;
	min-height: 200px;
	height: 100%;
	overflow-y: scroll;
}

.picked-color {
	width: 20px;
	height: 20px;
	margin: 5px;
}

.color-holder {
	display: flex;
	padding-top: 20px;
}

.palette-container {
	display: flex;
}

.delete-btn {
	border-radius: 5px;
	height: 25px;
	width: 100px;
	margin: 3px;
}
