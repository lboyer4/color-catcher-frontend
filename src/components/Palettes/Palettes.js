import React, { Component } from 'react';
import './_Palettes.scss';
import { connect } from 'react-redux';
import { addPalette, deletePalette } from '../../actions';
import PropTypes from 'prop-types';

export class Palettes extends Component {
	constructor() {
		super();
		this.state = {
			name: ''
		}
	}

	handleChange = (e) => {
		this.setState({name: e.target.value});
	};

	addPalette = () => {
		const colors = this.props.palette.reduce((acc, colorObject, index) => {
			acc[`color_${index+1}`] = colorObject.color.slice(1)
			return acc
		}, {})
		
		const newPalette = {...colors, name: this.state.name, project_id: this.props.project.id}
		this.postPalette(newPalette)
		}

	postPalette = (palette) => {
		let options = {
	    method: "POST",
	    headers: { "Content-Type": "application/json" },
	    body: JSON.stringify(palette)
    }
		fetch('http://localhost:3001/api/v1/palettes', options)
		.then(response => {
			if(!response.ok) {
				throw Error('Error posting palette')
			} else {
				return response
			}
		})
		.then(result => result.json())
		.then(body => this.props.addPalette({...palette, id: body.id}))
	}

		deletePalette = (e) => {
			let options = {
				method: "DELETE"
			};
			const paletteIdToDelete = e.target.parentElement.getAttribute('id')
			fetch(`http://localhost:3001/api/v1/palettes/${paletteIdToDelete}`, options)
			this.props.deletePalette(paletteIdToDelete)
		}


	render() {
		const matchingPalettes = this.props.palettes && this.props.palettes.filter(palette => {
			return palette.project_id === this.props.project.id
		});

		const createPaletteInput =
			<div> 
				<input className='palette-name' placeholder='Name your palette!' type='text' onChange={this.handleChange} value={this.state.name} />
				<button className='add-palette-btn' onClick={this.addPalette}>Add Palette </button>
			</div>

		const displayInput = this.props.project.name && createPaletteInput

		const displayPalettes = matchingPalettes && matchingPalettes.map(palette => {
				return (
					<div className='palette-container' key={palette.id}>
						<h2>{palette.name}</h2>
						<div className='color-holder' id={palette.id}>
						<div className='picked-color' style={{backgroundColor: `#${palette.color_1}`}}>
						</div>
						<div className='picked-color' style={{backgroundColor: `#${palette.color_2}`}}>
						</div>
						<div className='picked-color' style={{backgroundColor: `#${palette.color_3}`}}>
						</div>
						<div className='picked-color' style={{backgroundColor: `#${palette.color_4}`}}>
						</div>
						<div className='picked-color' style={{backgroundColor: `#${palette.color_5}`}}>
						</div>
						<button className = 'delete-btn' onClick={this.deletePalette}>Delete Palette</button>
						</div>
				</div>
				);
			});


		return (
			<div className='palette-holder'>
				<h1>{this.props.project.name}</h1>
				{displayInput}

				{displayPalettes}
			</div>
		);
	};
};

export const mapStateToProps = (state) => ({
	project: state.project,
	palette: state.palette,
	palettes: state.palettes
});

export const mapDispatchToProps = (dispatch) => ({
	addPalette: (palette) => dispatch(addPalette(palette)),
	deletePalette: (id) => dispatch(deletePalette(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(Palettes);

Palettes.propTypes = {
	addPalette: PropTypes.func,
	palette: PropTypes.array,
	palettes: PropTypes.array,
	project: PropTypes.object
}