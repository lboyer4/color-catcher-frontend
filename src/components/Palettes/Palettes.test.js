import React from 'react';
import { shallow } from 'enzyme';
import { Palettes, mapStateToProps, mapDispatchToProps } from './Palettes';
import { addPalette, deletePalette } from '../../actions';

describe('Palettes', () => {
	let wrapper;
	let mockProject;
	let mockPalette;
	let mockPalettes;
	let mockAddPalette;
	let mockDeletePalette;

	beforeEach(() => {
		mockAddPalette = jest.fn();
		mockDeletePalette = jest.fn();
		mockProject = {id: 29, name: "project"};
		let mockPalette = [
			{	locked: false,
				color: "#816379"
			},
			{ locked: false,
				color: "#816379"
			},
			{ locked: false,
				color: "#816379"
			},
			{	locked: false,
				color: "#816379"
			},
			{ locked: false,
				color: "#816379"
			}
		];
		mockPalettes = [
		{	color_1: "CF71BD",
			color_2: "D09B24",
			color_3: "0C2A28",
			color_4: "CC3EDC",
			color_5: "816379",
			name: "hi there",
			project_id: 30,
			id: 40},
		{	color_1: "CF71BB",
			color_2: "D09B22",
			color_3: "002A28",
			color_4: "CC3EDC",
			color_5: "886379",
			name: "bye there",
			project_id: 31,
			id: 41}
		];


		wrapper = shallow(
			<Palettes 
				addPalette={mockAddPalette}
				deletePalette={mockDeletePalette}
				project={mockProject}
				palette={mockPalette}
				palettes={mockPalettes}
			/>
		)
	});

	it('should match the snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});

	describe('handleChange', () => {
		it('should handle the change of name', () => {
			const mockNameEvent = {
				target: {
					value: 'Palette One'
				}
			};

			wrapper.instance().handleChange(mockNameEvent);
			expect(wrapper.state().name).toEqual('Palette One')
		});
	});

	describe('addPalette', () => {
		it('should call post palette', () => {
			const newPalette = {	
				color_1: "CF71BD",
				color_2: "D09B24",
				color_3: "0C2A28",
				color_4: "CC3EDC",
				color_5: "816379",
				name: "hi there",
				project_id: 30,
				id: 40
			}
			wrapper.instance().postPalette = jest.fn();
			wrapper.instance().addPalette();
			expect(wrapper.instance().postPalette).toHaveBeenCalled();
		});
	});

	describe('postPalette', () => {
		it.skip('should have a post fetch', () => {

		});
	});

	describe('deletePalette', () => {
		it.skip('should call a delete fetch', () => {

  
			})

		});
		it('should dispatch a delete palette action', () => {

			const paletteIdToDelete = 1;
			wrapper.instance().deletePalette = jest.fn();

			wrapper.instance().deletePalette = jest.fn()
			wrapper.instance().deletePalette()
			expect(wrapper.instance().deletePalette).toHaveBeenCalled()
		});
	

	describe('mapStateToProps', () => {
		it('should return a project object', () => {
			const mockState = {
				project: {
							name: "project one",
							id: 1
						},
				palette: [
			{	locked: false,
				color: "#816379"
			},
			{ locked: false,
				color: "#816379"
			},
			{ locked: false,
				color: "#816379"
			},
			{	locked: false,
				color: "#816379"
			},
			{ locked: false,
				color: "#816379"
			}
		],
		palettes: [
		{	color_1: "CF71BD",
			color_2: "D09B24",
			color_3: "0C2A28",
			color_4: "CC3EDC",
			color_5: "816379",
			name: "hi there",
			project_id: 30,
			id: 40},
		{	color_1: "CF71BB",
			color_2: "D09B22",
			color_3: "002A28",
			color_4: "CC3EDC",
			color_5: "886379",
			name: "bye there",
			project_id: 31,
			id: 41}
		]
			}
				
			const expected = {
				project: {
							name: "project one",
							id: 1
						},
				palette: [
			{	locked: false,
				color: "#816379"
			},
			{ locked: false,
				color: "#816379"
			},
			{ locked: false,
				color: "#816379"
			},
			{	locked: false,
				color: "#816379"
			},
			{ locked: false,
				color: "#816379"
			}
		],
		palettes: [
		{	color_1: "CF71BD",
			color_2: "D09B24",
			color_3: "0C2A28",
			color_4: "CC3EDC",
			color_5: "816379",
			name: "hi there",
			project_id: 30,
			id: 40},
		{	color_1: "CF71BB",
			color_2: "D09B22",
			color_3: "002A28",
			color_4: "CC3EDC",
			color_5: "886379",
			name: "bye there",
			project_id: 31,
			id: 41}
		]
			}

			const mappedProps = mapStateToProps(mockState)


			expect(mappedProps).toEqual(mockState)
		});
	});

	describe('mapDispatchToProps', () => {
		it('should call dispatch of addPalette', () => {
			const mockPalette = [
			{	locked: false,
				color: "#816379"
			},
			{ locked: false,
				color: "#816379"
			},
			{ locked: false,
				color: "#816379"
			},
			{	locked: false,
				color: "#816379"
			},
			{ locked: false,
				color: "#816379"
			}
		]
			const mockDispatch = jest.fn();
			const actionToDispatch = addPalette(mockPalette);

			const mappedProps = mapDispatchToProps(mockDispatch)

			mappedProps.addPalette(mockPalette)
			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
		});
		

		it('should call dispatch of deletePalette', () => {
			const mockId = 1
		const mockDispatch = jest.fn();
		const actionToDispatch = deletePalette(mockId);

		const mappedProps = mapDispatchToProps(mockDispatch)

		mappedProps.deletePalette(mockId)

		expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
		});
	});
});