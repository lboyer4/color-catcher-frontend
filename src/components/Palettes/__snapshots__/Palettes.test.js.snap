// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Palettes should match the snapshot 1`] = `
<div
  className="palette-holder"
>
  <div>
    <h2
      className="current-project"
    >
      Current Project: "
      <span>
        project
      </span>
      "
    </h2>
    <div
      className="input-holder"
    >
      <input
        className="palette-name"
        onChange={[Function]}
        placeholder="Name your palette!"
        type="text"
        value="Enter Palette Name Here!"
      />
      <button
        className="add-palette-btn"
        onClick={[Function]}
      >
        Add Palette 
      </button>
    </div>
  </div>
</div>
`;
